crates/txpool/src/pool.rs

```diff
@@ -0,0 +1,740 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use log::{trace, warn};
+use std::{
+    collections::{hash_map, BTreeSet, HashMap},
+    slice,
+    sync::Arc,
+};
+
+use crate::{
+    error,
+    listener::{Listener, NoopListener},
+    options::Options,
+    ready::{Readiness, Ready},
+    replace::{ReplaceTransaction, ShouldReplace},
+    scoring::{self, ScoreWithRef, Scoring},
+    status::{LightStatus, Status},
+    transactions::{AddResult, Transactions},
+    VerifiedTransaction,
+};
+
+/// Internal representation of transaction.
+///
+/// Includes unique insertion id that can be used for scoring explictly,
+/// but internally is used to resolve conflicts in case of equal scoring
+/// (newer transactionsa are preferred).
+#[derive(Debug)]
+pub struct Transaction<T> {
```

Comment:
We are creating a new type of transaction? :)

Reply:
Also, I would strongly advocate for having only one type of transaction and not parameterize this